design/specs/visual-query-builder.md

```diff
@@ -0,0 +1,236 @@
+
+Some options, such as filters and aggregations, may be available for columns that link to multiple records to help users define which data they want to show and how. The system will automatically detect the type of column and present users with the available options.
+
+### 2.1. Adding Input Columns
+
+To begin adding columns to a query configuration panel, the user must first select the `Add Column` button situated in the `Columns` portion of the query configuration panel. The `Column Selector` component will be displayed in the inspector panel as a result of this action.
+
+The `Column Selector` will display a list of all available columns as well as those already in use, if available. In addition, a complete list of all formulas will be available.
+
+The columns in `Column Selector` are listed in a hierarchical structure based on the links that exist in both directions between the base table and other tables. The icons indicate whether the linked records are a single or multiple record collection.
+
+Links are presented as expandable sections containing the columns from the linked tables.
+
+When adding input columns, the user can add them using the list controls or drop the columns directly into the result table.
+
+---
+Wireframes
+
+[Adding Columns](https://share.balsamiq.com/c/oyxTXxqSh8rLqU3JY71DWY.png)
+
+[Added Column](https://share.balsamiq.com/c/7U9yahZtYWX2G5obkyivoz.png)
+
+### 2.2 Input Column Sources
+
+Once the input columns have been added, the user can inspect them by selecting each from the result table or the list of columns in the query configuration panel, respectively. A section titled 'Source' will be included in the details for each column. The source column, table, and links will all be listed in this portion of the panel.
+
+The source will include `Links` only for data sourced via tables connected to or from the base table.
+
+---
+Wireframes
+
+[Column Sources](https://share.balsamiq.com/c/ewCxvkjS8VoScKfgb9hFos.png)
+
+### 2.3 Adding Formulas
+
+Formulas are used to generate new columns based on different parameters. To access the list of formulas, the user start the `Add Column` process and selects the option `From Formula` at the top of the inspector panel. Selecting a formula will open a form that users can fill out to determine the values of the new column.
+
+Depending on the selected formula, different settings will be available.
+More on formulas and specific details for each will be covered on a separate issue.
+
+Columns generated from formulas will display a formula icon indicator in the column header.
+
+---
+Wireframes
+
+[Added Formula](https://share.balsamiq.com/c/vWfJ9sUYWJxed5Zo5WtPGi.png)
+
+### 2.4 Filtering Input Column Values
+
+Columns that link to multiple records can have filters applied to them to retrieve only values that match user-specified criteria. Multiple filters are allowed for each input column.
```

Comment:
@mathemancer Could you provide an example of what you're envisioning here? I'm having a hard time understanding the issue from your description.

Reply:
Suppose we're aggregating actors in movies into lists, but we're only interested in actors whose first name starts with 'P'. Then filtering might produce some movies with _no_ actors after filtering for only actors fitting the description. I assume we'd want to show `NULL` for that case. What if instead, we're aggregating the number of Academy awards held by the casts of movies? This would be a sum, rather than a list. But then the value shown for a movie with a cast none of whom ever won an Oscar should probably be `0`, not `NULL`. Is this the intended UI in these cases? Are there other cases  to consider?

Now, since we're opening the possibility of ending up with _no_ results for some movies, what if we have some many-to-one relationship (maybe movie studios as a toy example)? Each movie has one studio, and for some reason we're only interested in studios starting with the letter 'P'. Why not give the possibility of filtering the linked studio column so that we show 'Pixar' and 'Paramount', but only `NULL` where there would have been 'Universal'? My question there is whether the restriction to only allowing filtering of linked records in cases of columns that link multiple records is needed.

Reply:
I understand your question now. I'm concerned that allowing filtering for those types of use cases would overload the concept of filters. I think users should use formulas for use cases like that instead, e.g. [the `Is True` formula](https://wiki.mathesar.org/en/product/specs/2022-01-views/04-formulas/4e-boolean-formulas#is-true).